README.md
# IS Publisher

> Provides access to information system backend..

## Installation

Clone repository, go to the cloned folder and then install all the package dependencies with the following command:
```sh
npm install
```

## Configuration

In order to properly configure the IS Publisher, create the following file from the root folder:
```sh
./config/instance.config.js
```

Below you can see how the structure of this file should be with the systems default values (you can omit the comments):

```javascript
module.exports = {
  "server": {
    "http": {
      //HTTP server port
      "port": 80
    }
  },
  "modules": {
    "infosystem": {
      "storage": {
        "_default": {
          "dialect": "couchdb",
          "options" :{
            "name": "_default",
            //The couchdb endpoint url.
            //You should provide the correct credentials.
            "url": "https://nouser:nouser@localhost:6984",
            //This is the default collection name.
            //Depends on ISCollector setup.
            "collection": "isappdb"
          }
        }
      },
      "rest": {
        //Location of GraphQL endpoint for
        //the REST API
        "graphQLUrl": {
          "protocol"  : "https",
          "host"      : "localhost",
          "port"      : 80,
          "path"      : "graphql"
        }
        //if you want to use the local graphql
        //service you can set
        //
        //  "graphQLUrl": "local"
        //
        //to avoid setting each field.
      }
    },
    "couchDBProxy": {
      //Location of the CouchDB endpoint to proxy
      "target": {
        "protocol"  : "https",
        "host"      : "localhost",
        "port"      : 6984,
        "username"  : "",
        "password"  : ""
      },
      //Path for CouchDB proxy to log traffic.
      //The process must have write permissions to the folder.
      "logpath"     : "/var/log/is-publisher/proxy"
    }
  }
};

```

## Starting the server

### Development
Execute from the root folder of IS Publisher the following command:
```sh
node bin/dev
```
### Production

Execute from the root folder of IS Publisher the following command:
```sh
npm run production
```

### Long running service

To ensure that the service will always be running there are two files, for production and development instances, that relies on [pm2 package](http://pm2.keymetrics.io/).

If pm2 is installed we can run for a development instance, the following command in the root folder of ispublisher:
```sh
pm2 start pm2.dev.json
```

For the production instance an extra step is required.
```sh
npm run build
pm2 start pm2.production.json
```

## Usage

### GraphQL interface

Provides a service endpoint to post [GraphQL](https://graphql.org/) queries. The service runs over HTTP via a single endpoint:
```
http://<ispublisher_hostname>/graphql
```

In order to help you build your queries or simply perform queries, the [graphiql](https://github.com/graphql/graphiql) UI tool is provided, at the following endpoint:
```
http://<ispublisher_hostname>/tools/graphiql
```
To visually inspect the information schema, the [voyager](https://github.com/APIs-guru/graphql-voyager)   web UI tool is also provided, at the following endpoint:
```
http://<ispublisher_hostname>/tools/voyager
```

>It is suggested to use the aforementioned tools since they provide the data schema documentation and can guide you to write correct queries.

### REST API interface
This service is build on top of graphql service to help users perform simple queries is case they cannot use the graphql language. The root path of this service is accessible at:
```
http://<ispublisher_hostname>/rest
```
The specific endpoint responds with a json describing all of the available REST resources in order to get started.

It should be noted that many complex query operations performed in graphql, such as quering array elements, or matching numeric ranges etc,do not apply to this interface. The same goes for the resource structure, which is predefined in the REST API case.

#### Paging with the REST API
Paging is achieved with the use of **limit** and **skip** query string fields. Eg. Retrieve the first 10 sites:
```
http://<ispublisher_hostname>/rest/sites?limit=10?skip=0
```

### Filtering with the REST API
Filtering is achieved by setting the **filter** query string field. It contains space seperated expressions in the following format:
```
  <property>[.<subproperty>][::<operation>]:<value>
```

| name | mandatory | description |
| ---- | --------- | ----------- |
| property | yes | The property of the resource |
| subproperty | no | A nested property |
| operation | no | A comparison operation. Defaults to **eq**(equals) |
| value | yes | The value to match |

>Properties and subproperties are case sensitive

>Text values must be wrapped in double quotes

>Boolean values take no operation and can be either **true** or **false**

Operation value depends on the value type:

| operation | value | description |
| --------- | ----- | ----------- |
| eq | text, number | Exact equal with value |
| ne | text, number | Not equal with value |
| like | text | Contains value. Can use "*". |
| ilike | text | Case insensite. Contains value. Can use "*". |
| nlike | text | Does not contain value. Can use "*" |
| nilike | text | Case insensitive. Does not contain value. Can use "*" |
| gt | number | Greater than value |
| ge | number | Greater than or equal to value |
| lt | number | Less than value |
| le | number | Less then or equal to value |


#### Examples
Find any site from _"Spain"_:
```
/rest/sites?filter=country:"Spain"
```

Find any site with name that contains the word _"meta"_ we write our filter like:
```
/rest/sites?filter=name::ilike:"*meta*"
```

Combine the above queries:
```
/rest/sites?filter=country:"Spain" name::ilike:"*meta*"
```

Find any site with a service not in _production_:
```
/rest/sites?filter=services.isInProduction:false
```

## CouchDB interface

This interface is a read only proxy to the backend CouchDB instance that can be used by couchdb client tools. It can be accessed from the following endpoint:
```
http://<ispublisher_hostname>/couchdb
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.2] - 2018-11-05
### Changed
- Update third party dependencies to their current version
- Use package.json version value from graphql version resolver
- Clean up unused code

## [1.3.1] - 2018-04-23
### Changed
- Code revision and documentation
- Revised logging information

## [1.3.0] - 2018-01-16
### Added
- Add WARNING in site service status enumeration
- Add enumeration filtering in REST api
- Add argo status and gocdb downtime in site and service details for REST api
- Add argo service statuses resource in REST api
- Add gocdb downtimes resource in REST api
### Changed
- Update apollo-server dependency package to v1.3.2
### Fixed
- Fix relation of service statuses to site services in entity model


## [1.2.1] - 2017-12-18
### Added
- Allow models to provide custom pre process arguments and post process response (future use)

## [1.2.0] - 2017-12-13
### Added
- Graphical tool voyager to display graphql schema
- Add filtering in REST api interface
- Add filtering of SRV and service downtimes from the sites and services entities in GraphQL
- Add cors in GraphQl endpoint
### Changed
- Grapgiql tool is moved under route /tools with voyager
- CouchDB optimization. Detect if query can be performed by id field only
- Removed "or" operations from filter types until properly implement backend functionality
- Auto convert of REST filters to GraphQL ones


## [1.1.3] - 2017-11-28
### Changed
- Set all filters to be handled woith filter operators instead of passing raw values
- CouchDB optimazation. Use request level caching to avoid duplicate db queries
- Ensure that only one db task is running any specific query and not having duplicate db requests
### Fixed
- Replaced invalid filter types for site service entity


## [1.1.2] - 2017-11-21
### Changed
- Set default couchdb collection to "isappdb"


## [1.1.1] - 2017-11-20
### Changed
- Set shared fields between entities.
- Relate service and SRV downtime entities with site and site service entities
- CouchDB optimization. Do not use $eq instead of $in operator for single id search
- CouchDB optimization. Merge external entity filtering with internal on shared fields
### Fixed
- Fix endpointURL mapping in SiteService model
- Fix graphql filtering for many to one relation types.


## [1.1.0] - 2017-11-15
### Added
- Added filter array operators (containsAll, icontainsAll, containsSome, icontainsSome, containsBetween)
- Added new fields in SiteService, SiteServiceImage, SiteServiceTemplate in Graphql and REST interface
- Added central logger
### Changed
- Set default production cluster to 4 process


## [1.0.0] - 2017-11-10
### Added
- initial release

config/config.js
var existsSync = require('fs').existsSync;
var _ = require('lodash');
var path = require('path');

var config = {
  "server": {
    "http": {
      "port": 80,
      "routes": {
        "infosystem":{
          "rest" : "/rest",
          "graphql": "/graphql",
          "graphiql": "/graphiql"
        },
        "couchDBProxy": "/couchdb"
      }
    }
  },
  "modules": {
    "infosystem": {
      "storage": {
        "_default": {
          "dialect": "couchdb",
          "options" :{
            "name": "_default",
            "url": "https://nouser:nouser@localhost:6984",
            "collection": "isappdb"
          },
          "logger": {
            "info": "/var/log/is-publisher/dbaccess/info.log",
            "debug": "/var/log/is-publisher/dbaccess/debug.log"
          }
        }
      },
      "graphql": {
        "logger": {
          "info": "/var/log/is-publisher/graphql/info.log",
          "debug": "/var/log/is-publisher/graphql/debug.log"
        }
      },
      "rest": {
        "graphQLUrl": "local",
        "logger": {
          "info": "/var/log/is-publisher/rest/info.log",
          "debug": "/var/log/is-publisher/rest/debug.log"
        }
      }
    },
    "couchDBProxy": {
      "target": {
        "protocol"  : "https",
        "host"      : "localhost",
        "port"      : 6984,
        "username"  : "",
        "password"  : ""
      },
      "logpath"     : "/var/log/is-publisher/proxy"
    }
  }
};


function loadConfig() {
  var _instanceConfigPath = path.resolve(__dirname, './instance.config.js');
  var instanceConfig = {};

  if (existsSync(_instanceConfigPath)) {
    instanceConfig = require(_instanceConfigPath);
  }

  return _.merge(config, instanceConfig);
};


module.exports = loadConfig();
